carbon-design-system/carbon-for-ibm-dotcom
# chore(masthead-v2): update main v2 branch with dev/L0 v2.1 changes

### Description

JIRA ticket: https://jsw.ibm.com/browse/ADCMS-2985

This PR should simply be getting `feat/masthead-v2` up to date with the changes from `feat/masthead-v2-dev`.


#### Why are there nearly 1000 associated commits?
I believe the reason is that in addition to integrating `feat/masthead-v2.1` feature work, `feat/masthead-dev` also recently merged in the latest from `main` ([8f9e0ba](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/commit/8f9e0bae3da5936da3c1e2f77b91c862d4d8f737)). Back in November 2022, we _also_ merged what was then the latest from `main` into `feat/masthead-v2`, but this was after `feat/masthead-v2-dev` already split off. This somehow caused git to be unable to find the two branches' common ancestor.

To fix this, I reverted the merge from 11/22, then merged `feat/masthead-v2-dev` into `feat/masthead-v2`, which applied cleanly.

### Testing Instructions

Please take a final, sweeping look at [the Masthead demo](https://ibmdotcom-webcomponents.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/10066/iframe.html?args=&knob-platform%20name%20(platform)=&knob-show%20the%20profile%20functionality%20(has-profile)=true&knob-show%20the%20search%20functionality%20(has-search)=true&knob-search%20placeholder%20(searchPlaceholder)=Search%20all%20of%20IBM&knob-selected%20menu%20item%20(selected-menu-item)=Consulting%20&%20Services=&knob-The%20user%20authenticated%20status%20(user-status)=Unauthenticated&knob-use%20mock%20nav%20data%20(use-mock)=true&id=components-masthead--with-v-2-data&viewMode=story) and verify it meets the [L0 v2.1 design & functional specs](https://www.figma.com/file/35JfN0Sg4ltIIjM7jBDKMk/Navigation?node-id=3179%3A173922&t=nXDyl46h5rLGsr4y-0). Note any issues you find on this PR. Thanks!

### Changelog

**New**

- Integrates latest changes from `feat/masthead-v2-dev`, including L0 v2.1 design updates.
